sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,802 @@
+JIT compiler.  This design document also describes how this SPIR-V
+transformation is done.
+
+
+## Properties on a global variable type
+
+One use for compile-time properties is with types that are used exclusively
+for declaring global variables.  One such example is the
+[SYCL\_EXT\_ONEAPI\_DEVICE\_GLOBAL][2] extension:
+
+[2]: <extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc>
+
+```
+namespace sycl::ext::oneapi {
+
+template <typename T, typename PropertyListT = property_list<>>
+class device_global {/*...*/};
+
+} // namespace sycl::ext::oneapi
+```
+
+The following code illustrates a `device_global` variable that is declared with
+two compile-time properties:
+
+```
+using sycl::ext::oneapi;
+
+device_global<int,
+  property_list_t<
+    device_image_scope::value_t,
+    host_access::value_t<host_access::access::read>>>
+  dm1;
+```
+
+The header file represents these properties with an internal C++ attribute
+named `[[__sycl_detail__::add_ir_global_variable_attributes()]]` whose value
+is a list that is created through a template parameter pack expansion:
+
+```
+namespace sycl::ext::oneapi {
+
+template <typename T, typename PropertyListT = property_list<>>
+class device_global {/*...*/};
+
+// Partial specialization to make PropertyListT visible as a parameter pack
+// of properties.
+template <typename T, typename ...Props>
+class
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_global_variable_attributes(
+    Props::meta_name..., Props::meta_value...
+    )]]
+#endif
+  device_global<T, property_list<Props...>> {/*...*/};
+
+} // namespace sycl::ext::oneapi
+```
+
+The `[[__sycl_detail__::add_ir_global_variable_attributes()]]` attribute has an
+even number of parameters.  The first half of the parameters are the names of
+the properties, and the second half of the parameters are the values for those
+properties.  Each property has exactly one value, so the property at parameter
+position 0 corresponds to the value at position _N / 2_, etc.  To illustrate
+using the same example as before, the result of the parameter pack expansion
+would look like this:
+
+```
+namespace sycl::ext::oneapi {
+
+template </* ... */> class
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_global_variable_attributes(
+    "sycl-device-image-scope",  // Name of first property
+    "sycl-host-access",         // Name of second property
+    nullptr,                    // First property has no parameter
+    "read"                      // Value of second property
+    )]]
+#endif
+  device_global</* ... */> {/*...*/};
+
+} // namespace sycl::ext::oneapi
+```
+
+The device compiler only uses the
+`[[__sycl_detail__::add_ir_global_variable_attributes()]]` attribute when the
+decorated type is used to create an [LLVM IR global variable][3] and the global
+variable's type is either:
+
+* The type that is decorated by the attribute, or
+* An array of the type that is decorated by the attribute.
```

Comment:
Should I read this as "we're adding a type attribute"?

I had originally assumed this was a declaration attribute based on the syntactic location of where it is written in the examples, but we aren't defining this in terms of the usual C++ language spec wording, so it's hard to tell what's intended.

Reply:
The example usage shows the `[[__sycl_detail__::add_ir_global_variable_attributes()]]` attribute in the syntactic location that applies the attribute to a class declaration, and a class declaration creates a user defined type.  I assumed that the attribute would apply to that user defined type.  Is there some other syntactic location we should use for this C++ attribute?  Sorry, I've been confused in the past about the syntactic locations of attributes, so maybe I made a mistake here.

Reply:
> Sorry, I've been confused in the past about the syntactic locations of attributes, so maybe I made a mistake here.

Yeah, this stuff is pretty confusing, so apologies definitely not necessary! The example shows it in the syntactic location of applying to the declaration of the type, not the type itself. So it technically only applies to the type when named specifically as the declared type (roughly) e.g., https://godbolt.org/z/4YP7583vv (note the GCC and MSVC behavior which are more closely related to how the standard suggests we should behave; the Clang behavior may be a useful bug).
```
struct [[decl_attr_1]] S { ... } [[type_attr]] name [[decl_attr_2]];
```
`decl_attr_1` is a declaration attribute for the declaration of `S`.
`type_attr` is a type attribute for the type `S`.
`decl_attr_2` is a declaration attribute for the variable named `name`.

Functionally speaking for Clang's implementation, `decl_attr_1` is attached to the AST node created for the structure `S`, so anything that looks at the canonical definition of the type and gets back to `S` can "see" the attribute. However, that information can be lost because it's not part of the type's identity. Consider: https://godbolt.org/z/zcej1Pas3 where we diagnose `S` but not `D`.

Clang does not have any type attributes that appertain to a struct, so we don't have much to model on there. Type attributes form a new type (or type qualifier) such as vector types, matrix types, address space qualifiers, nullability qualifiers, etc. so knowing whether you expect this to appertain to the type (as in, does it form a new type in the type system) or not will matter greatly in terms of implementation cost and behavior. So don't be afraid to ask more questions so that we can be sure all of us are on the same page!

Reply:
Reading between the lines, I think you are saying that it's much easier to implement the attribute in the position I have in the examples (declaration attribute) rather than in the position after the closing brace (type attribute).  Is that correct?

According to your godbolt examples, though, the only practical difference between the two **in clang** is the case when the decorated class is used as a base class.  In particular, a type alias created from the decorated type still behaves as though it has the decoration.  Correct?  I don't care about the behavior of the other compilers because this attribute will be used only in device code, and we always use our clang compiler for that code.

It is important in our use cases that a type alias created from the decorated type also behaves as though it is decorated with the attribute.  Technically, I think this means we want a "type attribute", but it seems like we get this behavior in clang even for a "declaration attribute".

Our immediate requirement for these attributes is for the new `device_global` class.  In that use case, we will never use `device_global` as a base class, so I think there is no immediately compelling reason to ask for a "type attribute" in the short term.

Longer term, we will want to use these attributes also for the `accessor` class, and that class can be used as a base class.  In this case, we would want the derived class to behave as though it was decorated with the attribute.  It sounds like we need to put the attribute in the "type attribute" position to get this behavior?

If my guess about it being easier to implement the "declaration attribute" is correct, then I think we could go with that in the short term.  Longer term, though, we might need to ask to support this attribute as a "type attribute".

Reply:
> An alternative would be to require the derived class to also specify the attributes of the base clas

This is not possible for the case I have in mind.  The `accessor` class is defined by us in our SYCL headers.  We do not plan to create derived types using `accessor` as a base.  However, our users might do this.  Since these attributes are an implementation detail, we can't ask our users to add the attribute to their derived type.

> But at the same time, this attribute is an implementation detail we're going to restrict to SYCL system headers (right?)

Yes, that's correct.  Moreover, our system headers ship together with the compiler.  Therefore, if we change the interpretation of these attributes later from "declaration attribute" to "type attribute", we can also change the headers that use them, and the two changes will be in sync.

Reply:
> This is not possible for the case I have in mind. The accessor class is defined by us in our SYCL headers. We do not plan to create derived types using accessor as a base. However, our users might do this. Since these attributes are an implementation detail, we can't ask our users to add the attribute to their derived type.

Yup, that sure sounds like a nonstarter now. Drat!

> Yes, that's correct. Moreover, our system headers ship together with the compiler. Therefore, if we change the interpretation of these attributes later from "declaration attribute" to "type attribute", we can also change the headers that use them, and the two changes will be in sync.

That depends entirely on how the attribute changes when becoming a type attribute and what requirements you have on your users. I can pretty easily see how this would create ABI issues for users because they silently get type mismatches across ABI boundaries, for example.

Reply:
If the "right" solution is to make these "type attributes", do we know how much extra work that is?  Maybe we should just do that in the first implementation.

Reply:
I *think* the right solution is to use a type attribute that modifies the type system (aka, adds a new `Type` node for the underlying type of declarations) but does not generate distinct types based on the arguments to the attribute (aka, the `Profile()` function for the type does not consider the arguments so each type is equivalent in terms of type checking). This should give you the behavior where you'd write:
```
struct whatever {
  ...
} [[sycl::attr(blah)]];

struct whatever {
  ... 
}; // error, redeclaration

struct whatever {
  ...
} [[sycl::attr(different_blah)]]; // error, also a redeclaration
```
but would give you the ability to do things like:
```
struct huttah {
  ...
} [[sycl::attr(blah)]];

typedef huttah buttery_goodness;

struct derived : huttah { 
  ...
};

// Using buttery_goodness will always track the [[sycl::attr(blah)]] bits
// Looking for type attributes on derived should also look for them on huttah
```
As for how much work is involved in this... I suspect we won't know until we get into it, but I would expect it to be fairly involved. Changing the type system is always exciting to begin with (we have to be careful not to negatively impact things like template instantiation depths when *not* using the attribute, etc). But also, this breaks new ground somewhat. For example, we have no type attributes that can be applied to a structure type, so I doubt we have any facilities for looking at attributed types through the structure's type hierarchy.

Reply:
FWIW, Erich and I are having an off-list conversation about this to make sure we're both on the same page. :-)

Reply:
Okay, we were not at all on the same pages it turns out. But we were reading the same book, so that's a good start. :-D

Erich reminded me of some problems with going with the type attribute approach and that may actually be a *worse* way to go right now (opposite of my earlier thinking). He's got me convinced to try the declaration attribute approach first with the understanding that we're likely to have somewhat of a long tail of bugs to find with it.

The primary issue with going with the type system approach I was thinking of is that I was incorrectly thinking of a mixture of `AttributedType` and a concrete type node, which won't work because of the design of `AttributedType` as being discardable type sugar. It's actually going to be *easier* to lose the attribute information my way because any step that "desugars" the type will drop the attributes on the floor. e.g., when doing member name lookup in Sema, we look to the type to find the record declaration and then look into that to find the member name, but to find the record declaration, we first strip all the type sugar off to get to a `RecordType`. Trying to find all of the places where we need to retain that sugar would be difficult, but making a wholly new `Type` node would be just as bad because now everywhere we need to look for a `RecordType` we would need to additionally say "or this other funky type that wraps a record type".

Because this information only needs to survive to codegen rather than to drive decisions within the FE, I think we will have an easier time of it going with a declaration attribute and manually handling the cases we need to as they come up (like when we want to inherit the property list from a base class onto a derived class).

Reply:
OK, so the resolution here is that we will keep the current syntactic location for these C++ attributes (declaration attributes).  I'll add a note, though, to capture the issue about inheriting the attribute's effect into derived classes, which is relevant for the `accessor` use case.

Reply:
Note that this means we would  NOT want to apply this to a typedef/type-alias, as these end up being useless there.  They need to be ONLY on the `RecordDecl` type, otherwise the 'meaning' of those is pretty novel.

Reply:
None of the C++ attributes proposed in this design will be applied to typedefs or type aliases, so I think this is fine.

Reply:
> None of the C++ attributes proposed in this design will be applied to typedefs or type aliases, so I think this is fine.

@elizabethandrews ^^ FYI!  Presumably this applies to the `uses_aspects` question you asked about.

Reply:
All this scary discussion on changing the type system reminds me the presentation from ARM at the ISO C++ meeting in Rapperswil when they wanted to change the C++ type system to fit their vector extension...
The goal here is to have this SYCL prototype up-streamed, right? :-)

Reply:
> I'll add a note, though, to capture the issue about inheriting the attribute's effect into derived classes, which is relevant for the accessor use case.

I did add this note in 4ec288158b945ecf68f098dd82a53e7a263b3014.

> All this scary discussion on changing the type system reminds me the presentation from ARM at the ISO C++ meeting in Rapperswil when they wanted to change the C++ type system to fit their vector extension...

After all that discussion we decided not to change the type system.